### ISSUE TITLE
(Re) Numeric Based Naming for Game Servers 

### ISSUE THREAD

**Issue by stevefan1999-personal** [agones-dev/agones#3905]
Previous discussion: #2286

My own reason for wanting to get this done: While Agones is clearly designed towards supporting lobby style matchmaking-based managed game server operator, but I want to have a stable, sorted ordinal that scales well with traditional game servers -- the one that is open to the public for hosting and usually community-backed, for example CS2, Rust ~~(the game, I do both Rust the language and Rust the game, how's the irony)~~, TF2 and Squad dedicated servers, and there's more, and the market for those games are still pretty high, yet often neglected in Agones anyway.

Because the players hosted their own community servers, other players get to choose which server to play from since you don't have a matchmaker, so if the users want to join the servers with their friends, it is not easy to do so. 

Right now, how I partially remediate this problem, is to read the Agones API and generate a server name suffix from it -- given that I use hash to make a small enough string like 5 random English characters and the chance of collision should be small and the players use this suffix code to find the right server -- while it works it is often quite confusing to them and if the server name is too long, then it is possible to spill out and the user couldn't see the server suffix or need to wait for the server name to roll like a carousel. Both problems can be addressed if we could just have a linear system that progressively rollout the game servers in a stable and ordinal way.

As such, why can't we just have it roll out like StatefulSet in Kubernetes that could generate a stable ordinal number so the players can easily recognize with. 

Oh, it can also serve as an indicator of how well the server scales as well.

**Comment by stevefan1999-personal** [comment]
no stale

**Comment by stevefan1999-personal** [comment]
no stale

**Comment by stevefan1999-personal** [comment]
no stale

**Comment by davidmayr** [comment]
I'd definitely vouch for this.

I have some thoughts to what @markmandel said about the problem with numeric ordering in the original issue https://github.com/googleforgames/agones/issues/2286#issuecomment-934912665:

Whilst I do definitely see his POV, and I agree that these numerical systems are far from perfect. Just like the original author of issue #2286 I'm also interested in such a numerical system for Minecraft as well. Mark's suggested safe words are a nice solution, but not what we are looking for. Also, these safe words would most likely need to be translated for international users, and then it's harder to share if two different people have different language settings. With the solutions I used before Agones we had noticed this downscale "problem" before, but we never considered it an issue at all, as it's just something you have to accept with such a naming system.

I don't even think players would give a second thought about it either if they see it in e.g. a lobby selection menu, or at least I would not as a player. Here is one of the larger German Minecraft servers (having at the time of making the screenshot 500 concurrent players), showing two lobbies in the selection menu, but the second lobby is named Lobby-3: https://imgur.com/a/P9LItYa. And although my Minecraft example probably does not fully represent the rest of the game dev world, I think the fact that this is so widely accepted in the Minecraft world should show that it's not really an issue to players.

The point I want to make is that people that would use this system are most likely well aware of the "downsides" and they are probably also aware why such a setting would not be the default. Unless there are some internal complications implementing this, I don't see a reason to not do this. Maybe there can even be more different naming schemes implemented in the future (like with the safe words, if someone wants it).

**Comment by markmandel** [comment]
An interesting way to potentially explore this space for an invested party, could be to create a third party controller based on just using `GameServer` instances, and see how it could work out -- since this is pretty big deviation from a standard Fleet, it probably needs to mostly be written from scratch anyway.

**Comment by stevefan1999-personal** [comment]
no-stale

**Comment by stevefan1999-personal** [comment]
no stale

**Comment by stevefan1999-personal** [comment]
no-stale

**Comment by stevefan1999-personal** [comment]
no-stale

**Comment by stevefan1999-personal** [comment]
no-stale

**Comment by markmandel** [comment]
> no-stale

@stevefan1999-personal if nothing else, I appreciate your commitment to these issues 😄

**Comment by stevefan1999-personal** [comment]
@markmandel because I need them so much in Agones internally rather than having to roll my own script. My solution is more like a dedicated server specific workaround, that I have to use some annotation hacks to do that, rather than a well-done solution.

**Comment by markmandel** [comment]
Appreciate that!

To be fair, as discussed, without a discussion of how to mitigate the issues particularly with scaling down, and then scaling up, not sure if we'll ever get to this in any capacity.

I'm always curious if a third party controller that does that lobby naming in a way that would help you out? (kinda what your script is doing?) - from what I'm reading it's about having a unique code for the lobby more than actual scaling in an ordered way?

**Comment by stevefan1999-personal** [comment]
> Appreciate that!
> 
> To be fair, as discussed, without a discussion of how to mitigate the issues particularly with scaling down, and then scaling up, not sure if we'll ever get to this in any capacity.
> 
> I'm always curious if a third party controller that does that lobby naming in a way that would help you out? (kinda what your script is doing?) - from what I'm reading it's about having a unique code for the lobby more than actual scaling in an ordered way?

I'm doing almost exactly that. I get a random hash from the host name, and turned it into a short mnemonic, and so the user associate the lobby they should go with that short mnemonic or I can even expose the hash as the server UID. Kind of hard to describe but that's the gist of it.

**Comment by markmandel** [comment]
So I guess my question is -- do you need real sequential ordering of game servers -- or do you just need a way to do some nicer naming on GameServer instances or their metadata?

Another fun idea for solving this problem would be a mutatingwebhookconfigurations that intercepts when GameServers are created, and does renaming for you within the cluster (or across clusters?) would need to be tied to a  DB of some kind and _fast_ would be good, since it's on the scaling hot path -- but it's another idea!

**Comment by stevefan1999-personal** [comment]
> So I guess my question is -- do you need real sequential ordering of game servers -- or do you just need a way to do some nicer naming on GameServer instances or their metadata?
> 
> Another fun idea for solving this problem would be a mutatingwebhookconfigurations that intercepts when GameServers are created, and does renaming for you within the cluster (or across clusters?) would need to be tied to a DB of some kind and _fast_ would be good, since it's on the scaling hot path -- but it's another idea!

Having a monotonic number that is larger than the last number is good enough. Having real sequential ordering of game servers (i.e. capable of reusing server \#2 as naming if the server was deallocated before) would be even better

**Comment by markmandel** [comment]
> Having a monotonic number that is larger than the last number is good enough. Having real sequential ordering of game servers (i.e. capable of reusing server [#2](https://github.com/googleforgames/agones/issues/2) as naming if the server was deallocated before) would be even better

Got it! That is definitely waaay simpler than having ordered scale up / scale down.

Does it have to be the GameServer name that is a number, or could it be an assigned label? 

The one advantage of having it be the GameServer name is that if you mess up and try and create one that already exists, it just won't work - k8s will stop you. So a third party mutation webhook would be a really good place to start. Label through a controller wouldn't get in the way of scaling though, and could be done after the GameServer is created and/or ready. So pros and cons.

Also - the number, unique per cluster, or unique across clusters?

Still not sure if I would advocate for this to be core Agones - but it definitely sounds like a solid third party project -- at least to experiment in to start.

**Comment by stevefan1999-personal** [comment]
> > Having a monotonic number that is larger than the last number is good enough. Having real sequential ordering of game servers (i.e. capable of reusing server [#2](https://github.com/googleforgames/agones/issues/2) as naming if the server was deallocated before) would be even better
> 
> Got it! That is definitely waaay simpler than having ordered scale up / scale down.
> 
> Does it have to be the GameServer name that is a number, or could it be an assigned label?
> 
> The one advantage of having it be the GameServer name is that if you mess up and try and create one that already exists, it just won't work - k8s will stop you. So a third party mutation webhook would be a really good place to start. Label through a controller wouldn't get in the way of scaling though, and could be done after the GameServer is created and/or ready. So pros and cons.
> 
> Also - the number, unique per cluster, or unique across clusters?
> 
> Still not sure if I would advocate for this to be core Agones - but it definitely sounds like a solid third party project -- at least to experiment in to start.

It could be both, but right now a number would be better. I have been relying on writing an external bash script for this

**Comment by stevefan1999-personal** [comment]
no-stale im actually going to work on it once my damn LLM finally refreshes its weekly usage on friday. My original sidecar operator has some breaking changes that makes it fragile.

**Comment by stevefan1999-personal** [comment]
no-stale

**Comment by stevefan1999-personal** [comment]
no-stale

**Comment by stevefan1999-personal** [comment]
no-stale
